insta/src/snapshot.rs

```diff
@@ -511,42 +596,107 @@ impl Snapshot {
         self.save_with_metadata(&new_path, &self.metadata)?;
         Ok(new_path)
     }
+
+    pub fn cleanup_extra_files(snapshot: Option<&Self>, path: &Path) -> Result<(), Box<dyn Error>> {
+        let binary_path = snapshot.and_then(|snapshot| match &snapshot.contents() {
+            SnapshotContents::Text(_) => None,
+            SnapshotContents::Binary(contents) => Some(contents.build_path(path)),
+        });
+        let binary_file_name = binary_path.as_ref().map(|path| path.file_name().unwrap());
+
+        let file_name = path.file_name().unwrap();
+
+        let read_dir = path.parent().unwrap().read_dir();
+
+        if read_dir
+            .as_ref()
+            .is_err_and(|e| e.kind() == ErrorKind::NotFound)
+        {
+            return Ok(());
+        }
+
+        for entry in read_dir? {
+            let entry = entry?;
+            let entry_file_name = entry.file_name();
+            if entry_file_name
+                .as_encoded_bytes()
+                .starts_with(file_name.as_encoded_bytes())
+                && entry_file_name != file_name
+                && binary_file_name.map_or(true, |x| x != entry_file_name)
+            {
+                std::fs::remove_file(entry.path())?;
+            }
+        }
+
+        Ok(())
+    }
 }
 
 /// The contents of a Snapshot
-// Could be Cow, but I think limited savings
 #[derive(Debug, Clone)]
-pub struct SnapshotContents {
+pub enum SnapshotContents {
+    Text(TextSnapshotContents),
+    Binary(BinarySnapshotContents),
+}
+
+// Could be Cow, but I think limited savings
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct TextSnapshotContents {
     contents: String,
-    pub kind: SnapshotKind,
+    pub kind: TextSnapshotKind,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct BinarySnapshotContents {
```

Comment:
We have a lot of new data structures! I'm thinking whether it's possible to reduce how many we need. As an example — is this needed, or can a binary snapshot just store the bytes? The only thing this seems to do is to build its path, which arguably the binary snapshot should do already?

Reply:
I don't quite know what you meant with that last part. Because a `Snapshot` doesn't know its own path, that only gets passed in from context. Which also means that we don't need to mutate it when approving for example. This also extends to the binary snapshots, which build the paths on demand based on the relevant pending or stored path. In the #489 implementation I had the path in the `SnapshotContents` and that made some things a little annoying and weird because we needed to update that in the right places and so on.

But anyway, we do already have the extension in the `MetaData` struct which is redundant. So I've changed it to use the extension from there. I suppose the whole thing is not quite optimal because we don't have the typesystem enforcing that the snapshot type in the metadata and in the contents is the same, but enforcing that would be a pretty big change I think because it would likely mean not having the `metadata` field in `Snapshot` struct, but like everything from metadata except for the exension and then generating the full metadata at the moment of saving or something like that. So I think assuming them to be the same thing plus asserting and unwrapping in some places is an acceptable trade-off.

Reply:
I very much agree with your second paragraph, I've been thinking around the same lines. There's a decent amount of duplication in the data structures, which is a smell.

Possibly there's some big refactor we could do in the future (should we have `SnapshotAssertionContextFile` given lots there is specific? `SnapshotAssertionContext<Kind>`?? How to generalize across FileText & Binary; vs FileText & InlineText?). But I don't think we have to compress everything at this point, as long as the code is nicely legible.

Thanks for trying the change in https://github.com/mitsuhiko/insta/pull/610/commits/1b4f7ec471fc50356a222bf7b24dce3c9f5b4722. After seeing it, I think it's possibly slightly better, though not as much of a simplification as I had imagined, partly because we can't specialize on the Binary vs Text at the Snapshot struct. Am fine either way if you want to keep / revert.